### PR TITLE
New post about TAOCP

### PR DESCRIPTION
**Pull request by adamrmelnyk**

### CODE REVIEW

File: content/post/the-art-of-computer-programming.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
++++
+title = "The Art of Computer Programming"
+date = 2018-08-25T17:56:57-05:00
+description = "Making my way through TAOCP"
+draft = false
+toc = false
+categories = ["computer science"]
+tags = ["taocp", "knuth"]
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: Would a "vol1" tag make sense?

File: content/post/the-art-of-computer-programming.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+images = [] # overrides the site-wide open graph image
++++
+
+An attempt at making my way through one of the most infamous computer science books
+
+I was recently gifted the first volume of Donald Knuth's: The Art of Computer Programming. Being computer geek that I am, I have never been so excited yet, intimidated about sitting down and working through what some people would call a Reference book.
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: Title should be in quotes.

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: >Being _the_ computer geek...

File: content/post/the-art-of-computer-programming.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+<!--more-->
+
+## Reading the book
+
+The biggest misconception that I had about this book was that I thought it would be more like a reference book, something you can quickly open, flip to the table of contents, and then over to page 254 for linked lists (or linked allocation in this case) and learn everything there is to know about linked lists. As stated by peers, stack overflow posts, reddit, and hacker news users, it's not something you would want to actually go about reading through. It's long, highly technical, and full of math equations. I suspect that the average reader would quickly get lost the first time Knuth shows off an Algorithm in a custom assembly language written specifically for this book (yes, really). To put this more plainly: <b>Reading through TAOCP is an NP hard problem</b>. Naturally, I couldn't wait to read through it. So how should we go about reading the book? Luckily for us Knuth provides the reader with instructions, and of course, being the computer scientist that he is, a helpful algorithm flow chart to guide the reader through the process of reading the books which looks something like this:
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: Capitalize them websites.

File: content/post/the-art-of-computer-programming.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+     +-----+------+
+           |
+           |
+  +--------v---------+
+  |4. Begin Chapter N|
+  +--------+---------+
+           |
+           |
+   +-------v-------+ Yes       +---------------------------------------+
+   |5. Interesting?+---------->| 6. Read until Exaustion or Whatever   |
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: >Exhaustion

File: content/post/the-art-of-computer-programming.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+   +---------------+           | the CS equivalent of Nirvana might be |
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: lolololol

File: content/post/the-art-of-computer-programming.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+           |No                 +---------------------------------------+
+           |                                        ^
+           |                                        |
+           +----------------------------------------+
+```
+
+The flow chart and written proceedure following this contains other suggestions on how to read through the book based on if you've read through it before or just based on what your interests are.
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: > procedure

File: content/post/the-art-of-computer-programming.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+### Exercises
+
+A large portion of this book is dedicated to exercises as well as the included soultions in the back. <b><i>Ugh, homework</i></b>. One of the more helpful things that Knuth does is break down the relative difficulty of each problem using this notation:
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: >solutions

File: content/post/the-art-of-computer-programming.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+```
+00 Immedidiate                        > Recommended
+10 Simple (one minute)                M Mathematically oriented
+20 Medium (15 minutes)                HM Requiring "higher math"
+30 Moderately Hard (You might)
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: "You might" be able figure it out?

File: content/post/the-art-of-computer-programming.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+40 Term Project 
+50 Research problem
+```
+
+In his reasoning for breaking down excercises this way, Knuth includes an anecdote about how Richard Bellmans book <i>Dynamic Programming</i> mixed trivial problems with research problems with no indication about which problems were solvable or not. <b><i>I bet his students loved that</i></b>. Though I wanted to get a lot out of this book the first time round I don't expect to go through all of these. Probably the easier ones to get the idea of it, but also doing a little bit of searching about the more Research type problems is certainly interesting.
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: > exercises

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: > Bellman's

File: content/post/the-art-of-computer-programming.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
...engthy disection of what algorithms are as well as the history behind the word Algorithm, Section 1.2 includes a lot of the foundational math behind computer sciences, inductions, permutations and combinatorics, harmonic numbers, and elementary number theory; All the things you tried to wrap your head around in Algorithms I throught III that you thought you could forget about have come back with a vengeance. I'll admit that when I started reading the introduction to this chapter and flipped through the next 10 pages or so I considered picking up some supplimentary materials on the math involved. After some searching I came across some of his other works and I almost picked up his book on [concrete mathematics](https://www.amazon.com/dp/0201558025/). It was at this point that my adventures in computer science had started to look a lot like [this](https://www.youtube.com/watch?v=AbSehcT19u0). I resisted the temptation a bit and pressed on which I'm glad I did because two paragraphs down Knuth suggests you lightly skim this the first time you read through this otherwise you might never get around to reading the actual computer science topics. <b><i>Hmmm... yeah that checks out</i></b>
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: > dissection

File: content/post/the-art-of-computer-programming.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+## MIX
+
+Not only did knuth have to [create a typesetting language after the second edition of the second volume was created](https://en.wikipedia.org/wiki/TeX#History) but he also set about creating an assembly language for a made up machine called MIX, reasoning that progamming languages tend to influence the way that an engineer might code so having a more pure language to work with might solve that issue. <i>[Hmm... That sounds familiar](https://www.youtube.com/watch?v=AbSehcT19u0)</i>. You can find the entire set of instructions and and diagram of the machine architecture [here](https://en.wikipedia.org/wiki/MIX) although if you can find a copy of the book I recommend looking at the diagram given in the book itself because it's beautiful. What I found more interesting is that this architecture is actually supported by [gcc](https://gcc.gnu.org/onlinedocs/gcc/MMIX-Options.html) meaning we can compile code for the theoretical computer. Knuth describes this comuter, the MIX 1009 that runs MIX assembly code as:
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: >Knuth have to

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: >instructions and diagram

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: >this computer,

File: content/post/the-art-of-computer-programming.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+<i>MIX is very much like nearly every computer of the 1960's and 70's. except that it is, perhaps, nicer.</i>
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: >70's,

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: Consider block-quoting.

File: content/post/the-art-of-computer-programming.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+There you have it. Throw out that new macbook pro you bought for $3000 and <i>rm -rf</i> that Vue JS project you've been working on because it's time we threw something new into the MIX! Knuth does admit that at this point MIX is a bit antequated so naturually he notes that a new version, dubbed [MMIX](http://mmix.cs.hm.edu/index.html), based on RISC, has been created and will likely be included in subsequent editions of the book, just as soon as they rewrite all of the algorithms in MMMIX assembly.
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: Capitalize that beastly machine's name.

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: Consider `<code>` tag or backticks for that `rm -rf`.

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: >something new into the MIX!

lolololol

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: > antiquated

Comment (chrisbodhi) [reply content/post/the-art-of-computer-programming.md]: >~~M~~ MMIX

File: content/post/the-art-of-computer-programming.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+## Concluding thoughts
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: >Thoughts

File: content/post/the-art-of-computer-programming.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+At this point we might as well say it. Completing this book seems to be an impossible task, before Knuth can document everything, previous parts of it have already started to go out of date and the various <i>"Under construction"</i> sections throughout the book highlight just how quickly, and how much information there really is about this field of study, though that's not particularly that surprising to anyone who's spent time trying to include the newest npm module into their React project. I spent the better part of 4 years studying computer science in school yet barely scratched the surface of what the field has to offer and years later I'm still finding the need to learn more about it. At times it can feel overwhelming just trying to keep up with learning the basics of the newest language, the hottest Javascript framework or the the newest trends. Software engineering, web development, and comuter sciences are moving at blazingly fast speeds with nary an end in sight.
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: >and computer sciences are moving

File: content/post/the-art-of-computer-programming.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+So why read this then? Though the book is quite old I still think that there is a lot of value in reading through it. Much of it is still relevant today as data structures and basic algorithms (the underpinning of everything we still use daily) have not changed much, and more than that; Knuth is actually a fantastic writer. Though the material may be heavy, Knuth has a sort of dry humor that I've come to appreciate that has made reading this volume enjoyable. Though I don't think that [reading TAOCP alone will get you a job](https://www.businessinsider.com/bill-gates-loves-donald-knuth-the-art-of-computer-programming-2016-4), I recommend reading TAOCP for anyone who wishes to pique their intellectual curiousities about the field of computer science.
```
Comment (chrisbodhi) [inline content/post/the-art-of-computer-programming.md]: > curiosities

File: public/categories/computer-science/index.html [hunk 17]
```diff
@@ -0,0 +1,267 @@
+<!DOCTYPE html>
+<html lang="en-US">
+  <head>
+    <meta name="generator" content="Hugo 0.47.1" />
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Computer Science | Adams&#39; Blag</title>
```
Comment (chrisbodhi) [inline public/categories/computer-science/index.html]: I think the setting for this is elsewhere, but the apostrophe in the ~~blog~~ blag title is in the wrong place, under more than one Adam is contributing.

File: content/post/the-art-of-computer-programming.md (hunk 2 above)
Comment (adamrmelnyk) [reply content/post/the-art-of-computer-programming.md]: good catch

File: public/categories/computer-science/index.html (hunk 17 above)
Comment (adamrmelnyk) [reply public/categories/computer-science/index.html]: yup good catch

**Pull request merged**